### PR TITLE
feat: Add azuredevops_group_entitlment

### PR DESCRIPTION
**Pull request by stanleyz**
This change is based on PR #707, most of the work should be credited to marley-ma.

https://github.com/microsoft/terraform-provider-azuredevops/pull/707

"

## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

Issue Number:

https://github.com/microsoft/terraform-provider-azuredevops/issues/700

Issue Number: 700

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Any relevant logs, error output, etc?

(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -223,5 +223,6 @@ The following Environment Variables must be set in your shell prior to running a
 - `AZDO_DOCKERREGISTRY_SERVICE_CONNECTION_USERNAME`
 - `AZDO_GITHUB_SERVICE_CONNECTION_PAT`
 - `AZDO_TEST_AAD_USER_EMAIL`
+- `AZDO_TEST_AAD_GROUP_ID`
```
Comment (xuzhang3) [inline README.md:226]: `AZDO_TEST_AAD_GROUP_ID`  is for `azuredevops_group_entitlment` test, no need to document in this file

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+		Delete: resourceGroupEntitlementDelete,
+		Update: resourceGroupEntitlementUpdate,
+		Importer: &schema.ResourceImporter{
+			State: importGroupEntitlement,
+		},
+		Schema: map[string]*schema.Schema{
+			"principal_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Optional: true,
```
Comment (xuzhang3) [inline azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]: `principal_name`  cannot be updated
```suggestion
				Computed: true,
```

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+		"origin_id",
+	}
+)
+
+func ResourceGroupEntitlement() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGroupEntitlementCreate,
+		Read:   resourceGroupEntitlementRead,
+		Delete: resourceGroupEntitlementDelete,
+		Update: resourceGroupEntitlementUpdate,
```
Comment (xuzhang3) [inline azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]:
```suggestion
		Create: resourceGroupEntitlementCreate,
		Read:   resourceGroupEntitlementRead,
		Update: resourceGroupEntitlementUpdate,
		Delete: resourceGroupEntitlementDelete,
```

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+	if err != nil {
+		return fmt.Errorf("Creating group entitlement: %v", err)
+	}
+
+	addedGroupEntitlement, err := addGroupEntitlement(clients, groupEntitlement)
+	if err != nil {
+		return fmt.Errorf("Creating group entitlement: %v", err)
+	}
+
+	flattenGroupEntitlement(d, addedGroupEntitlement)
```
Comment (xuzhang3) [inline azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]:
```suggestion
	d.SetId(addedGroupEntitlement.Id.String())
```

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+func resourceGroupEntitlementRead(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+	groupEntitlementID := d.Id()
+	id, err := uuid.Parse(groupEntitlementID)
+	if err != nil {
+		return fmt.Errorf("Error parsing GroupEntitlementID: %s. %v", groupEntitlementID, err)
+	}
+	groupEntitlement, err := readGroupEntitlement(clients, &id)
```
Comment (xuzhang3) [inline azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]:
```suggestion
		groupEntitlement, err := clients.MemberEntitleManagementClient.GetGroupEntitlement(clients.Ctx, memberentitlementmanagement.GetGroupEntitlementArgs{
		GroupId: &id,
	})

```

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+	result := *patchResponse.Results
+
+	if !*result[0].IsSuccess {
+		return fmt.Errorf("Updating group entitlement: %s", getGroupEntitlementAPIErrorMessage(&result))
+	}
+	return resourceGroupEntitlementRead(d, m)
+}
+
+var uuidRegexp = regexp.MustCompile("^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$")
+
```
Comment (xuzhang3) [inline azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]:
```suggestion

```

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+func importGroupEntitlement(d *schema.ResourceData, m interface{}) ([]*schema.ResourceData, error) {
+	_, err := uuid.Parse(d.Id())
+	if err != nil {
+		upn := d.Id()
+		if !uuidRegexp.MatchString(upn) {
+			return nil, fmt.Errorf("Only UUID values can used for import [%s]", upn)
+		}
```
Comment (xuzhang3) [inline azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]:
```suggestion
		
```

**Review by xuzhang3 — CHANGES_REQUESTED** [review]
@stanleyz  A few changes requested, others looks good to me. 
Document `website/docs/r/group_entitlement.html.markdown` need to be documented in `azuredevops.erb`

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go (hunk 5 above)
Comment (stanleyz) [reply azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]: this is same as the function call, isn't it?

File: README.md (hunk 1 above)
Comment (stanleyz) [reply README.md:226]: This is to follow the pattern of  `AZDO_TEST_AAD_USER_EMAIL` which is used by `azuredevops_user_entitlement` test and a few others. If not documented here, anywhere else you'd recommend as a better place?

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go (hunk 7 above)
Comment (stanleyz) [reply azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]: This is still required, uuid.parse doesn't valid the same as this regex. See the function `TestGroupEntitlement_Import_TestInvalidValue` in [resource_group_entitlement_test.go](https://github.com/stanleyz/terraform-provider-azuredevops/blob/114735f2953d6161a016f95b7e723a77d4ad26fb/azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement_test.go#L289)

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go (hunk 6 above)
Comment (stanleyz) [reply azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]: see below response

**Comment by stanleyz** [comment]
> @stanleyz A few changes requested, others looks good to me. Document `website/docs/r/group_entitlement.html.markdown` need to be documented in `azuredevops.erb`

@xuzhang3 I've added the link to the `azuredevops.erb`, also commented on the ones that haven't been resolved, can you please check? Thanks.

File: README.md (hunk 1 above)
Comment (xuzhang3) [reply README.md:226]: No need to document this as this is used for AccTest.  The pre-check in test will throw an error if not set

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go (hunk 5 above)
Comment (xuzhang3) [reply azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]: Yes. The get operation only called once so there is no need to wrap it into a function

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go (hunk 7 above)
Comment (xuzhang3) [reply azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]: The `d.id()`  is from service which we can trust is a valid UUID. If the ID is not a valid UUID, there is a something wrong with the service

File: README.md (hunk 1 above)
Comment (stanleyz) [reply README.md:226]: Well, the section it's documented to is called `## Environment variables for acceptance tests`, plus the `AZDO_TEST_AAD_USER_EMAIL` is also for AccTest only. I agree however README.md might not be a good place for such doc, but I'd suggest we keep `AZDO_TEST_AAD_GROUP_ID` here for consistency and move this whole section to a better place, for example `docs/contributing.md` sounds good for such docs.

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go (hunk 7 above)
Comment (stanleyz) [reply azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]: normally yes, however this function is for importing existing resource, you can't guarantee the users will always have a valid UUID supplied, it's not a big deal but additional check with nice error message always helps.

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go (hunk 5 above)
Comment (stanleyz) [reply azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]: Yeah, I realize that, all good, different coding style, no biggie, I've changed that.

File: azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go (hunk 7 above)
Comment (xuzhang3) [reply azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]: UUID already check by [_, err := uuid.Parse(d.Id())](https://github.com/microsoft/terraform-provider-azuredevops/pull/870/files#diff-301e91e27f45058b37c7d3f701322ee24e17cf3ed4ba81fa81e310ea83fbc6bdR250), no need to verify twice

Comment (stanleyz) [reply azuredevops/internal/service/memberentitlementmanagement/resource_group_entitlement.go]: cool, updated. That was my impression in the first look, double checked, there was a logic error in the test, not the `uuid.Parse`. I've updated the test and import function.

**Comment by xuzhang3** [comment]
```log
=== RUN   TestAccGroupEntitlement_Create
=== PAUSE TestAccGroupEntitlement_Create
=== RUN   TestAccGroupEntitlement_AAD_Create
=== PAUSE TestAccGroupEntitlement_AAD_Create
=== CONT  TestAccGroupEntitlement_Create
=== CONT  TestAccGroupEntitlement_AAD_Create
--- PASS: TestAccGroupEntitlement_AAD_Create (32.29s)
--- PASS: TestAccGroupEntitlement_Create (32.78s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        38.848s

```

**Pull request merged**